File: Bit Manipulation/Python/Single Number II.py
# Problem link: https://leetcode.com/problems/single-number-ii/description/
# Time: O(n)
# Space: O(1)
class Solution:
    def singleNumber(self, nums: List[int]) -> int:
        store = [0] * 32
        result = 0

        for num in nums:
            for pos in range(32):
                if num & (1 << pos):
                    store[pos] += 1
        
        for index in range(32):
            store[index] %= 3
            if store[index]:
                result += (2 ** index)

        return result if result < (1 << 31) else result - (1 << 32)

File: Bit Manipulation/Java/Single Number II.java
// Problem link: https://leetcode.com/problems/single-number-ii/description/
// Time: O(n)
// Space: O(1)
class Solution {
    public int singleNumber(int[] nums) {
        int[] store = new int[32];
        int result = 0;

        for(int num : nums) {
            for(int pos = 0; pos < 32; pos++) {
                if((num & (1 << pos)) != 0) {
                    store[pos] += 1;
                }
            }
        }

        for(int pos = 0; pos < 32; pos++) {
            store[pos] %= 3;
            if(store[pos] == 1) {
                result |= (1 << pos);
            }
        }

        return (result & (1 << 31)) == 0 ? result : result - (1 << 32);
    }
}
